### ISSUE TITLE
Router Dialyzer Bug When Upgrading to 1.4.15

### ISSUE THREAD

**Issue by asummers** [phoenixframework/phoenix#3697]
### Environment

* Elixir version (elixir -v):
```
Erlang/OTP 22 [erts-10.4.4] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.10.0 (compiled with Erlang/OTP 21)
```
* Phoenix version (mix deps): `1.4.15`
* NodeJS version (node -v): N/A
* NPM version (npm -v): N/A
* Operating system: OSX

### Expected behavior
When upgrading from `1.4.15` from `1.4.14` the router is now emitting a Dialyzer error in downstream `router.ex` files which is exposing a parsing bug in `Erlex` and then `Dialyxir` accordingly. 

```erlang
{:warn_matching, {'lib/your_app/router.ex', 1}, {:pattern_match, ['pattern <_@1, _@2 = \'true\'>', '<<<_:8,_:_*1>>,\'false\'>']}}
```

See discussion in https://elixirforum.com/t/dialyzer-issue-with-phoenix-1-4-15/29687/6

### Actual behavior

The router file no longer causes Dialyzer errors for folks.

**Comment by chrismccord** [comment]
To be clear, this is an Erlex bug or Dialyxir bug, or both? Or is this something you think is caused by Phoenix. If someone would like to take a look here it would be greatly appreciated.

**Comment by asummers** [comment]
The bug is in Phoenix 1.4.5 and not in 1.4.4 but it exposed some bad bugs in the other libraries. If anyone is using them they will get unintelligible errors because of the Phoenix bug.

**Comment by chrismccord** [comment]
Just before myself or someone else deep dives on this – you are positive this is a phoenix spec issue? Or is it possible our latest features trigger a bug in these other libraries?

**Comment by asummers** [comment]
I am positive that the issue originated somebwere between 1.4.4 and 1.4.5 though don't know the addition that caused it yet. I don't know if the error is an incorrect spec or a new addition violating an existing spec.

Simply updating caused the error to be exposed in my case as well as the person linked in the thread. I don't know offhand if it's broken on a mix phx.new but am willing to test any new branch to assert the changes work in my codebase.

**Comment by asummers** [comment]
Also those bugs never originate Dialyzer warnings, they simply reformat the existing warnings to be human readable so I'm 100% sure it's not the fault of either Dialyxir or Erlex.

**Comment by josevalim** [comment]
To be clear, there are three bugs here:

1. Dialyzer is surfacing an error using core syntax, instead of writing it in Erlang term syntax
2. Phoenix v1.4.15 has a dialyzer bug - that dialyzer does not report properly due to the above
3. erlex/dialyxir is not expecting Dialyzer to behave like 1, which then crashes

All three bugs should allegedly be fixed. The most important to be fixed is the erlex one, as the behaviour seen in `1` is not exclusive to Phoenix. Phoenix maintainers have historically not fixed dialyzer warnings, since we don't use them ourselves, so contributions are welcome here.

**Comment by nurugger07** [comment]
@josevalim experienced this yesterday & I've been looking at it this morning. Curious enough setting the default value in the Scopes struct `trailing_slashes?: true` gets a "correct" warning message from dialyzer. My opinion is, it's not really a Phoenix issue but more a dialyzer issue.

This also excludes `ErLex` I think because it's really just calling `:parser.parse/1` with what dialyzer sends over. I've run into "odd" dialyzer issues before.

**Comment by josevalim** [comment]
In this case, it was a Phoenix bug. If you didn't use any trailing_slashes, which is the default for all Phoenix apps before this release, we would generate a clause like this:

    def maybe_append_slash(path, true), do: ...

dialyzer knew this clause would never be triggered, because none of the calls to it has trailing slashes enabled, and then warn. The fix is simply to not emit this clause if there are no trailing slashes. A simpler fix would be to mark those functions clauses as generated, but skipping the clause altogether is better in runtime (less checks).

**Comment by mcrumm** [comment]
🤦‍♂In hindsight, my mistake is very clear. Sorry for all the trouble, and thanks for the assist @josevalim!

**Comment by nurugger07** [comment]
@josevalim yeah, I think there's a couple things happening here. Agree there is something in phoenix for the warning but what `:dialyzer.run/1` is returning seem to be "off." I won't keep responding here unless you want to track it for others but essentially the warnings coming back from dialyzer are slightly different and one is causing an error when parsing the message.

When the value is set to true, you get a proper warning. Working great and that should alert the user that there's an issue to resolve, in this case in phoenix.
Return from dialyzer:
```
[
  {:warn_matching, {'lib/elixir_fountain_web/router.ex', 1},
   {:pattern_match_cov, ['pattern <_@1, _>', '<<<_:8>>,\'true\'>']}}
]
```
When the value is set to false, it's unable to parse the message from dialyzer and `:parser.parse/1` throws an error from ErLex
Return from dialyzer:
```
[
  {:warn_matching, {'lib/elixir_fountain_web/router.ex', 1},
   {:pattern_match, ['pattern <_@1, _@2 = \'true\'>', '<<<_:8>>,\'false\'>']}}
]
```
Note the difference in `:pattern_match_cov` vs `:pattern_match`. 

As mentioned, I won't continue on here unless you want me to :) Trying to locate an fix in the other libraries that are causing the underlying issue.

**Comment by coladarci** [comment]
Sorry to comment on an old issue, but I was about to file this exact same issue. I am having a hard time understanding what the resolution is. Bumping erlex got rid of the parse error for me, but I'm still stuck here:
```
lib/web/router.ex:1:pattern_match
The pattern can never match the type.
Pattern:
_, _ = true
Type:
<<_::8, _::size(1)>>, false
```

```
Erlang/OTP 22 [erts-10.6.4] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]
Elixir 1.10.2 (compiled with Erlang/OTP 22)
phoenix 1.4.15
dialyxir 1.0.0-rc.7
erlex 0.2.6
```

Is this an issue on my end, in that I did something stupid in my router?

As always, thanks!

**Comment by josevalim** [comment]
It is a Phoenix bug that has been fixed in the repo. Use Phoenix directly from the v1.4 git branch until a new release is out.

**Comment by ffimnsr** [comment]
The fix for the emit trailing slash is not on v1.4 branch but on master. 

https://github.com/phoenixframework/phoenix/commit/b48bd80918ebbffe3b33a4c57be49703a1f806ae

**Comment by shahryarjb** [comment]
there are some bugs on router phoenix 1.4.16, umbrella project
```
lib/phoenix/router.ex:2:no_return
Function call/2 has no local return.
________________________________________________________________________________
lib/phoenix/router.ex:2:no_return
Function call/2 has no local return.
________________________________________________________________________________
lib/phoenix/router.ex:331:pattern_match_cov
The pattern
variable_

can never match, because previous clauses completely cover the type
:error.

________________________________________________________________________________
lib/phoenix/router.ex:331:pattern_match_cov
The pattern
variable_

can never match, because previous clauses completely cover the type
:error.

________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```

will it fixed on new phoenix like 1.4.17 ?

**Comment by josevalim** [comment]
We don't run dialyzers ourselves, so please send a PR if you want to see those fixed in an upcoming release.
